Dmay/job launcher refactor

psanan opened the pull request.

dmay23 committed:
Job: Slight modifications of class
- Added empty method clean()
- Added walltime as optional (kwargs) parameter. Arguably walltime could/should be included in the resource dict().
- Added createJobOrdering() method for JobSequence class
- Added resource type "idle_mpi_ranks_per_node". Rational for "idle" versus "active". We need to compute the "max" resources required by a job. As such recording idle makes more sense than active. It is an awkward choice though for the user to set.

dmay23 committed:
Launcher updates: Class members remain unchanged, methods and implementations have been altered. Launcher now accepts Job as an argument (and not Test). Major changes in the implementations are largely required to support JobSequence and JobDAG jobs.
- Launcher.clean() only removes files it creates
- os.system() has been replaced with the subprocess module. Currently this commit is only compatible with python3
- Support for LLQ has been removed (there are no BGQ machines in usage anymore).
- This commit DOES not execute the call to launch batch jobs (still in testing phase)

dmay23 committed:
Launcher test: Added examples using Job, JobDAG

dmay23 committed:
test/launcher/test1: Removed out-of-date keyword arg

dmay23 committed:
test/launcher/test1: updated keyword argname "walltime" -> "wallTime"

dmay23 committed:
Py2/3 subprocess module: Adds a naive wrapper for subprocess which supports both python flavours

dmay23 committed:
job.py: Change exit(1) --> Raise()

dmay23 committed:
Job: Bug fix in JobSequence when returning order names

dmay23 committed:
Job: Add method to return list of jobs (in order they will be executed)
This was added as the job sequence cannot get jobs efficiently based on their textual name (e.g. from the return result of job.createJobOrdering()

dmay23 committed:
Launcher: slight refactor associated with error code collection
- Major change is that all error codes (from sequence/DAG job) are shoved into a single file. This enables test.verify() to easily determine and report that a job may have failed due to failure of a dependent job.
- Centralised how standard output files (error code,stdout,stderr) are named.

dmay23 committed:
Wall time variables: Cleanup usage of wall time variable
* Walltime is passed to launcher, it is a value computed by the job
* Job - added method to sum wall time required for a job, and any dependent jobs
* Update test1.py and removed the wall time kwarg variable from launcher

dmay23 committed:
JobDAG: Update comment for getJobList

dmay23 committed:
launcher: removed dead code

dmay23 committed:
Test: Added new test implementation
* Added standard status status codes (and colours for pretty printing)
* Added new verifier class
* Added unix diff verifier class

dmay23 committed:
UnixDiff: Fixed output_path usage

dmay23 committed:
Test test.py: Added unit test for test class and verifier and verifierUnixDiff

dmay23 committed:
Test: added verify() method
* Fixed indentation

dmay23 committed:
Test unit test: Added example of staged execute / verify.
* Improved usage of output_path

dmay23 committed:
Test unit test: Fixed indentation

dmay23 committed:
Job: Improved error checking in DAG
- Fixed comment

dmay23 committed:
Job composite test: Added test for composite jobs defined by either sequence+sequence, or dag+sequence

dmay23 committed:
Job getMaxWallTime bug fix: Correctly sum the wall time when dealing with composite jobs

dmay23 committed:
Job: Indentation change only

psanan commented:
I want to make sure that the existing examples (which are part of the current tests) continue to work. Here's a proposed plan to do that:

1. Add a `VerifierLegacy` class (in its own file, to be removed once obsolete)
2. Move all the `compareXYZ()` functions which to be class methods of `VerifierLegacy`
3. Update the examples, hence tests, to use this (expecting to update them to use new-and-improved approaches later). If some example is particularly laborious to update this way, I might just remove it.

[Update: not doing this, too much pointless work]

psanan commented:
I think it's a waste of time to make the existing harness work, before refactoring it, so I'll restrict my ambitions here to simply add the new tests to the CI, taking the tests of the existing examples out for now, to be re-added as the Harness is refactored.

dmay23 commented:
Yes! Delete harness and start again. And thus disable old tests for the
time being. That’s what we agreed upon when we spoke last week. When I
reimplement smartDiff I’ll put the old (non harness) tests back in.

On Thu 31. Oct 2019 at 14:54, Patrick Sanan <notifications@github.com>
wrote:

> I think it's a waste of time to make the existing harness work, before
> refactoring it, so I'll restrict my ambitions here to simply add the new
> tests to the CI, taking the tests of the existing examples out for now, to
> be re-added as the Harness is refactored.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sciath/sciath/pull/47?email_source=notifications&email_token=ADO6RJ2RZPKHFJ5PNMF3SNLQRLPSTA5CNFSM4JF2WJNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECX3NHY#issuecomment-548386463>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADO6RJ6FXPSY4GA35KKBS5DQRLPSTANCNFSM4JF2WJNA>
> .
>

psanan committed:
Remove outdated tests and examples

psanan committed:
Self-tests: update Job tests

psanan commented on sciath/launcher.py:
```diff
@@ -48,135 +48,268 @@ def formatMPILaunchCommand(mpiLaunch,ranks,corespernode):
+
+    return errorCodeName, stdoutName, stderrName
+
+def _removeFile(file2rm):
```
Is it safe to turn off "safety mode" here? I am trying to add your new tests to the self-tests before proceeding, but they need to run without user input.

psanan replied:
(For now, working around this by doing e.g. `yes | ./runTests.py`)

psanan committed:
Self-tests: add current Launcher and Test tests

Splits the Test tests up and does not test the unixdiff variants
left in tests/test/unittest_ex2.py

Note that the launcher testing requires confirmation of deleting files,
so in the CI we supply "yes |"

dmay23 commented:
Yes, safe to turn this off. It’s primarily intended for safety when
developing debugging the harness / sandbox-ing where there is lots of path
manipulations.

Maybe make safety mode a global module variable. As per any other debugging
flags. They are very useful and avoid trashing something by mistake.

On Thu 31. Oct 2019 at 15:32, Patrick Sanan <notifications@github.com>
wrote:

> *@psanan* commented on this pull request.
> ------------------------------
>
> In sciath/launcher.py
> <https://github.com/sciath/sciath/pull/47#discussion_r341171291>:
>
> > +    stderrName = []
> +
> +    lc_count = len(jobnames)
> +    for i in range(0,len(jobnames)):
> +        jprefix = "".join(["sciath.depjob-",str(lc_count),'-',jobnames[i]])
> +        if lc_count == 1: # we do something special for the last job in a sequence/DAG list
> +            jprefix = "sciath.job-" + job.name
> +
> +        stdoutName.append( jprefix + ".stdout" )
> +        stderrName.append( jprefix + ".stderr" )
> +
> +        lc_count -= 1
> +
> +    return errorCodeName, stdoutName, stderrName
> +
> +def _removeFile(file2rm):
>
> Is it safe to turn off "safety mode" here? I am trying to add your new
> tests to the self-tests before proceeding, but they need to run without
> user input.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/sciath/sciath/pull/47?email_source=notifications&email_token=ADO6RJ3B4P7AOSEWNPAQIS3QRLUBNA5CNFSM4JF2WJNKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCJ4RIFA#pullrequestreview-309924884>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADO6RJ5FOUHH76SN4SGUOS3QRLUBNANCNFSM4JF2WJNA>
> .
>

psanan committed:
Test/Verifier: pass path argument when verifying, instead of storing in Test

This makes Test/Verifier much more "functional" (no state or side effects) and easier to reason about and maintain.

When the Harness is refactored, it will be responsible for determining the paths where tests should be run from.

psanan committed:
Launcher: enable batch submission with old-style os call

Tested on Euler (LSF) with
```
import os

from sciath.job import Job
from sciath.launcher import Launcher

job_launcher = Launcher()
job_launcher.configure()
job_launcher.setVerbosityLevel(1000)

jA = Job(['echo','random-text-from-example_1'],name='DMDA
interpolation')
jA.setResources(ranks=4,threads=1)

path = './sandbox'

if not os.path.exists(path):
      os.mkdir(path)

      job_launcher.submitJob(jA,path=path)
      #job_launcher.clean(jA,path=path)

```

psanan commented:
@dmay23 I'd like to merge this, since it's already huge, and work on remaining fixes and improvements in new PRs to dev.

Note that this PR totally breaks compatibility with pyTestHarness, but I'm not bothering to bump the version since the other open PR does that.

dmay23 commented:
Why did you use os.system()? Use the internal py2/3 compatible function
please.

On Thu 31. Oct 2019 at 17:46, Patrick Sanan <notifications@github.com>
wrote:

> @dmay23 <https://github.com/dmay23> I'd like to merge this, since it's
> already huge, and work on remaining fixes and improvements in new PRs to
> dev.
>
> Note that this PR totally breaks compatibility with pyTestHarness, but I'm
> not bothering to bump the version since the other open PR does that.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/sciath/sciath/pull/47?email_source=notifications&email_token=ADO6RJ33IN7GESTWT7PTH6DQRMDW5A5CNFSM4JF2WJNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECYOZMA#issuecomment-548465840>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADO6RJ5I4RN75WIDRTT4JVTQRMDW5ANCNFSM4JF2WJNA>
> .
>

psanan commented:
Because I tried it and it failed, and you had it commented out so I figured it wasn't trivial, and I wanted something running that we could iterate on in other, contained, issues (opened issue #50 for this). Is it going to work when the command has a redirect in it? (e.g. `bsub < foo.lsf`).
```
Created submission file: ./sandbox/sciath.job-DMDA_interpolation-launch.lsf
[Executing DMDA_interpolation]
  [cmd]  ['bsub < ', './sandbox/sciath.job-DMDA_interpolation-launch.lsf']
Traceback (most recent call last):
  File "./test3.py", line 19, in <module>
    job_launcher.submitJob(jA,path=path)
  File "/cluster/home/pasanan/code/sciath/sciath/launcher.py", line 652, in submitJob
    ctx = subp.run(launchCmd)
  File "/cluster/apps/python/3.6.0/x86_64/lib64/python3.6/subprocess.py", line 403, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/cluster/apps/python/3.6.0/x86_64/lib64/python3.6/subprocess.py", line 707, in __init__
    restore_signals, start_new_session)
  File "/cluster/apps/python/3.6.0/x86_64/lib64/python3.6/subprocess.py", line 1326, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: 'bsub < '
```

dmay23 commented:
Sure sure, but let’s not put the code which is lame and which we want
removed back in - rather let’s get the improved approach working (note I’ve
removed os.system() completely from Test, Job, Launcher).

Recall I said “code not at all tested on any batch system”.  So I never
tested the commented line either. :(

Can you try this:
cmd = [‘bsub’, ’<‘,  ’foo.lsf’]
In general there should be no white space.



On Thu 31. Oct 2019 at 18:19, Patrick Sanan <notifications@github.com>
wrote:

> Because I tried it and it failed, and you had it commented out so I
> figured it wasn't trivial, and I wanted something running that we could
> iterate on in other, contained, issues. Is it going to work when the
> command has a redirect in it? (e.g. bsub < foo.lsf).
>
> Created submission file: ./sandbox/sciath.job-DMDA_interpolation-launch.lsf
> [Executing DMDA_interpolation]
>   [cmd]  ['bsub < ', './sandbox/sciath.job-DMDA_interpolation-launch.lsf']
> Traceback (most recent call last):
>   File "./test3.py", line 19, in <module>
>     job_launcher.submitJob(jA,path=path)
>   File "/cluster/home/pasanan/code/sciath/sciath/launcher.py", line 652, in submitJob
>     ctx = subp.run(launchCmd)
>   File "/cluster/apps/python/3.6.0/x86_64/lib64/python3.6/subprocess.py", line 403, in run
>     with Popen(*popenargs, **kwargs) as process:
>   File "/cluster/apps/python/3.6.0/x86_64/lib64/python3.6/subprocess.py", line 707, in __init__
>     restore_signals, start_new_session)
>   File "/cluster/apps/python/3.6.0/x86_64/lib64/python3.6/subprocess.py", line 1326, in _execute_child
>     raise child_exception_type(errno_num, err_msg)
> FileNotFoundError: [Errno 2] No such file or directory: 'bsub < '
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/sciath/sciath/pull/47?email_source=notifications&email_token=ADO6RJ4DYVPEA6YK7HN5LOLQRMHTRA5CNFSM4JF2WJNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECYSOWQ#issuecomment-548480858>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADO6RJ7O7CVCDKQIWXWF2W3QRMHTRANCNFSM4JF2WJNA>
> .
>

psanan commented:
I don't think that works directly
```
$ python
Python 3.7.4 (default, Sep  7 2019, 19:52:29)
[Clang 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess
>>> subprocess.run(["cat","test.txt"])
foo
CompletedProcess(args=['cat', 'test.txt'], returncode=0)
>>> subprocess.run(["cat","<","test.txt"])
cat: <: No such file or directory
foo
CompletedProcess(args=['cat', '<', 'test.txt'], returncode=1)
```
(Also see e.g. https://stackoverflow.com/questions/3679974/run-shell-command-with-input-redirections-from-python-2-4)

dmay23 commented:
Ok. So read the file contents and shove it into stdin (see second answer)

On Fri 1. Nov 2019 at 12:05, Patrick Sanan <notifications@github.com> wrote:

> I don't think that works directly
>
> $ python
> Python 3.7.4 (default, Sep  7 2019, 19:52:29)
> [Clang 10.0.1 (clang-1001.0.46.4)] on darwin
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import subprocess
> >>> subprocess.run(["cat","test.txt"])
> foo
> CompletedProcess(args=['cat', 'test.txt'], returncode=0)
> >>> subprocess.run(["cat","<","test.txt"])
> cat: <: No such file or directory
> foo
> CompletedProcess(args=['cat', '<', 'test.txt'], returncode=1)
>
> (Also see e.g.
> https://stackoverflow.com/questions/3679974/run-shell-command-with-input-redirections-from-python-2-4
> )
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/sciath/sciath/pull/47?email_source=notifications&email_token=ADO6RJZIT6BVKMDRBMJJYK3QRQENZA5CNFSM4JF2WJNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEC2UJWQ#issuecomment-548750554>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADO6RJZYJ3UOXGZHM6UEQVDQRQENZANCNFSM4JF2WJNA>
> .
>

psanan commented:
What do you think about this idea? We add some more logic to our custom `run()` function, so that in addition to working with python 2 and 3, it can (optionally, on by default) also interpret some small subset of the bash redirection symbols. In particular (and this is all we'd initially implement), it would look for an entry in the command list equal to `"<"`, in the second-to-last position, and if found, use the last argument to open a file and send it as stdin to a function call corresponding to the remaining arguments. This would cover the `bsub <` case (which would become the `['bsub','<']` case).

psanan commented:
To be more concrete, and for reference when I come back to this, I mean something like this:
```
def run(command,stdoutFile,stderrFile,interpret_redirect_symbols=True) :
  """ Command akin to subprocess.run(), but with both Python 2 and Python 3
      support, and with an additional option to specially interpret
      a unix-style redirect "<" in the second-to-last slot """
  if stdoutFile is None:
      stdoutFile = subp.PIPE
  if stderrFile is None:
      stderrFile = subp.PIPE

  if interpret_redirect_symbols and len(command) >= 3 and command[-2] == '<':
      stdinFile = open(command[-1],'r')
      command = command[:-2]
  else:
      stdinFile = None

  try:
      if sys.version_info[0] == 2 :
          errorno = subp.call(command,stdin=stdinFile,stdout=stdoutFile,stderr=stderrFile)
      else :
          ctx = subp.run(command,universal_newlines=True,stdin=stdinFile,stdout=stdoutFile,stderr=stderrFile)
          errorno = ctx.returncode
  finally:
      if stdinFile:
          stdinFile.close()

  return errorno
```

dmay23 commented:
On Tue, 5 Nov 2019 at 07:28, Patrick Sanan <notifications@github.com> wrote:

> What do you think about this idea? We add some more logic to our custom
> run() function, so that in addition to working with python 2 and 3, it
> can (optionally, on by default) also interpret some small subset of the
> bash redirection symbols. In particular (and this is all we'd initially
> implement), it would look for an entry in the command list equal to "<",
> in the second-to-last position, and if found, use the last argument to open
> a file and send it as stdin to a function call corresponding to the
> remaining arguments. This would cover the bsub < case (which would become
> the ['bsub','<'] case).
>

Sounds okay to me.... but....
There is only one line in the entire code base which needs < .
So rather do something complex like changing the behaviour of what happens
when the second last item in the list of commands was "<", for this one
line in sciath where < is required I would preferably opt to just open the
file and jam it in to subp as stdin...


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sciath/sciath/pull/47?email_source=notifications&email_token=ADO6RJ65IE5GRYTRUTCKW73QSEODNA5CNFSM4JF2WJNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDB3SHA#issuecomment-549697820>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADO6RJ566A47OCNYW3YCHK3QSEODNANCNFSM4JF2WJNA>
> .
>

psanan commented:
It's not precisely "one line"  - the launcher will somehow have to store information about this redirect, which is annoying in the first place because it breaks the convenience assumption that a "command" is a list of strings [exec, arg, arg, .. ].  I'm not sure that handling this will be any less complexity! (Though you can argue that this approach could cause confusion in that we only handle a subset of things you can do in bash - I could just add some error messages for now if you have a ">", "|", or a "<" in any position but the second-to-last).


> Am 07.11.2019 um 12:31 schrieb dmay23 <notifications@github.com>:
> 
> On Tue, 5 Nov 2019 at 07:28, Patrick Sanan <notifications@github.com> wrote:
> 
> > What do you think about this idea? We add some more logic to our custom
> > run() function, so that in addition to working with python 2 and 3, it
> > can (optionally, on by default) also interpret some small subset of the
> > bash redirection symbols. In particular (and this is all we'd initially
> > implement), it would look for an entry in the command list equal to "<",
> > in the second-to-last position, and if found, use the last argument to open
> > a file and send it as stdin to a function call corresponding to the
> > remaining arguments. This would cover the bsub < case (which would become
> > the ['bsub','<'] case).
> >
> 
> Sounds okay to me.... but....
> There is only one line in the entire code base which needs < .
> So rather do something complex like changing the behaviour of what happens
> when the second last item in the list of commands was "<", for this one
> line in sciath where < is required I would preferably opt to just open the
> file and jam it in to subp as stdin...
> 
> 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/sciath/sciath/pull/47?email_source=notifications&email_token=ADO6RJ65IE5GRYTRUTCKW73QSEODNA5CNFSM4JF2WJNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDB3SHA#issuecomment-549697820>,
> > or unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/ADO6RJ566A47OCNYW3YCHK3QSEODNANCNFSM4JF2WJNA>
> > .
> >
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub <https://github.com/sciath/sciath/pull/47?email_source=notifications&email_token=AB5HBL6AYJNSZ6OB7VA53XTQSP4CJA5CNFSM4JF2WJNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDMDQVQ#issuecomment-551041110>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB5HBL2HGG7JQSM37VXIRP3QSP4CJANCNFSM4JF2WJNA>.
>

psanan committed:
Launcher: use sh to interpret < in bsub command

psanan commented:
Pushed what I think is a nice, non-invasive option. Instead of `bsub <`, which `subprocess.run()` can't handle because it doesn't interpret `<`, just pass through `sh` which can run a string as commands, handling the redirects, and accept arguments, i.e. `sh -c 'bsub < $0'` (Note single quotes). This seems to work fine and keeps everything simple.
